compiler/rustc_middle/src/mir/syntax.rs

```diff
@@ -237,19 +237,19 @@ pub enum StatementKind<'tcx> {
 
     /// `StorageLive` and `StorageDead` statements mark the live range of a local.
     ///
-    /// Using a local before a `StorageLive` or after a `StorageDead` is not well-formed. These
-    /// statements are not required. If the entire MIR body contains no `StorageLive`/`StorageDead`
-    /// statements for a particular local, the local is always considered live.
-    ///
-    /// More precisely, the MIR validator currently does a `MaybeStorageLiveLocals` analysis to
-    /// check validity of each use of a local. I believe this is equivalent to requiring for every
-    /// use of a local, there exist at least one path from the root to that use that contains a
-    /// `StorageLive` more recently than a `StorageDead`.
-    ///
-    /// **Needs clarification**: Is it permitted to have two `StorageLive`s without an intervening
-    /// `StorageDead`? Two `StorageDead`s without an intervening `StorageLive`? LLVM says poison,
-    /// yes. If the answer to any of these is "no," is breaking that rule UB or is it an error to
-    /// have a path in the CFG that might do this?
+    /// At any point during the execution of a function, each local is either allocated or
+    /// unallocated. Except as noted below, all locals except function parameters are initially
+    /// unallocated. `StorageLive` statements cause memory to be allocated for the local while
+    /// `StorageDead` statements cause the memory to be freed. Using a local in any way (not only
+    /// reading/writing from it) while it is unallocated is UB.
+    ///
+    /// Some locals have no `StorageLive` or `StorageDead` statements within the entire MIR body.
+    /// These locals are implicitly allocated for the full duration of the function. There is a
+    /// convenience method at `rustc_mir_dataflow::storage::always_storage_live_locals` for
+    /// computing these locals.
+    ///
+    /// If the local is already allocated, calling `StorageLive` again is UB. However, for an
+    /// unallocated local an additional `StorageDead` all is simply a nop.
```

Comment:
I wonder if we should consider double-free to also be UB, just for consistency? The only reason I allow this is that we do generate redundant StorageDead in MIR currently. (Or at least we did last time I checked this.)

Reply:
Recently reported #98896 indicates this is still an issue.

Reply:
I was actually going to suggest that we make the `StorageLive` version of this not be UB either. I don't see what we gain from adding the UB

Reply:
How would you suggest translating StorageLive to LLVM ir if double StorageLive isn't UB? I believe `StorageLive(_1); _1 = 42; StorageLive(_1); _2 = _1;` is UB in LLVM ir.

Reply:
That should be UB in MIR too - the semantics I am suggesting for `StorageLive` are that calling it when the local is already live is allowed but still fully reallocates the local

Reply:
I did this mostly out of paranoia because I think LLVM might consider this to be UB
```
StorageLive(_1)
Use(_1)
StorageLive(_1)
```
It looks like LLVM has changed their wording in the LangRef, it no longer says that uses before a `lifetime.start` are UB. However it also doesn't explicitly allow consecutive `start`. And it says

> After ‘llvm.lifetime.start’, the stack object that ptr points is marked as alive and has an uninitialized value. 

So at the very least, we'd have to overwrite the contents of the local with `Uninit` on a redundant `StorageLive`.

Reply:
> allowed but still fully reallocates the local

Oh I see. Yes that would satisfy the "reset to uninit" condition.

FWIW a separate `StorageReallocate` has been suggested in the past for this purpose (https://github.com/rust-lang/rust/issues/61849, https://github.com/rust-lang/rust/issues/68622).

Reply:
Still we should get confirmation from LLVM that `lifetime.start` will not have ant effects that flow "backwards" in the execution and make preceding accesses UB -- ideally, to explicitly state in their docs "doing this on an already live local is fine (but reset its contents to be uninitialized)".

Reply:
> If ptr is a non-stack-allocated object, it does not point to the first byte of the object or it is a stack object that is already alive, it simply fills all bytes of the object with poison.

<https://llvm.org/docs/LangRef.html#llvm-lifetime-start-intrinsic>

Seems like they already do

Reply:
Oh, I didn't see the "already alive" part. Good point. And for `lifetime.end` it says

> Calling llvm.lifetime.end on an already dead alloca is no-op.

So yeah this removes my main motivation for making redundant annotations UB in the first place. That means I do not have a strong opinion either way. I don't know if there are any good motivations for doing one or the other, or any optimizations/transformations that would be helped by either choice.